### PR TITLE
[AOTI] Introduce an extensibility mechanism for the c shim codegen to make it easy to produce c shims for out-of-tree OP kernels as well. Add c_shim for XPU.

### PR DESCRIPTION
**Pull request by etaf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #135320
* #135318
* #139026
* __->__ #136742
* #139025

[AOTI] Introduce an extensibility mechanism for the c shim codegen to make it easy to produce c shims for out-of-tree OP kernels as well. Add c shim for XPU.

### Motivation
Since the current c shim codegen will only produce C wrappers for Op's registered in `aten/src/ATen/native/native_functions.yaml`, for the same backend, when a portion of out-of-tree OP's are not registered in that file, but are registered externally. For example, `third_party/torch-xpu-ops/yaml/native_functions.yaml` , in this case, the existing codegen can't fulfill the need to do extensions for the c shims from the out-of-tree OPs for the in-tree that has already been produced.

### Design
To extend the c shim with more OP for a backend from out-of-tree.
The PR provided a bool option `--aoti-extend` to indicate the codegen is to extend c shim from out-of-tree.
The generated c shim is stored in the `extend` subdirectory , for example:
```
torch/include/torch/csrc/inductor/aoti_torch/generated/c_shim_xpu.h
torch/include/torch/csrc/inductor/aoti_torch/generated/c_shim_xpu.cpp
torch/include/torch/csrc/inductor/aoti_torch/generated/extend/c_shim_xpu.h
torch/include/torch/csrc/inductor/aoti_torch/generated/extend/c_shim_xpu.cpp
```
example usage:
`python -m torchgen.gen --source-path third_party/torch-xpu-ops/yaml/ --xpu --aoti-extend --update-aoti-c-shim  ` 
`--xpu`:  generate c shim for XPU
`--aoti-extend `: this is an out-of-tree OPs(defined in `third_party/torch-xpu-ops/yaml/native_functions.yaml`)  extend for in-tree ops(defined in `aten/src/ATen/native/native_functions.yaml`)
`--update-aoti-c-shim`: always generate c_shim_xpu.h for the extend c_shim.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torchgen/gen.py [hunk 1]
```diff
@@ -2850,12 +2858,15 @@ def main() -> None:
         help="Update AOTInductor C shim after adding an entry to inductor_fallback_ops in torchgen/aoti/fallback_ops.py. "
         "WARNING: Do not use this unless you are sure what you are doing!!!",
     )
+    parser.add_argument(
+        "--aoti-extend",
+        action="store_true",
+        help="Update AOTInductor C shim after adding an entry to inductor_fallback_ops in torchgen/aoti/fallback_ops.py. "
```
Comment (EikanWang) [inline torchgen/gen.py:2864]: Update the help message.

File: torchgen/gen_aoti_c_shim.py [hunk 2]
```diff
@@ -329,18 +330,24 @@ def get_backend_index_for_aoti(
...
-        backend_index = backend_indices[DispatchKey.CompositeExplicitAutograd]
-    elif backend_indices[DispatchKey.CompositeExplicitAutogradNonFunctional].has_kernel(
-        func
-    ):
-        # We need to create C shim wrappers for CompositeExplicitAutogradNonFunctional kernels
-        backend_index = backend_indices[
+    else:
+        # for the extend out-of-tree kernels, we don't need to
+        # duplicatly create C shim wrappers for other dispatch keys
+        if aoti_extend:
```
Comment (EikanWang) [inline torchgen/gen_aoti_c_shim.py:336]:
```suggestion
        if aoti_extend and diapatch_key == "XPU":
```

**Review by EikanWang — APPROVED** [review]

File: torchgen/gen.py [hunk 3]
```diff
@@ -2355,7 +2354,16 @@ def operator_headers() -> list[str]:
                     structured_func_group_dict[func.structured_delegate] = func_group
                     break
 
-        if dispatch_key in (DispatchKey.CPU, DispatchKey.CUDA):
+        aoti_c_shim_backends = [DispatchKey.CPU, DispatchKey.CUDA]
+        # generate aoti c shim for external backends
+        ext_backend_yamls = {
+            DispatchKey.XPU: "third_party/torch-xpu-ops/yaml/xpu_functions.yaml"
```
[the remark this answers is not in the document]
Comment (etaf) [reply torchgen/gen.py]: Thanks @malfet , I've updated the code.

File: torchgen/gen_aoti_c_shim.py (hunk 2 above)
Comment (etaf) [reply torchgen/gen_aoti_c_shim.py]: > 
The flag `aoti_extend`, which now is refined as `extend_aoti_c_shim`, is by design used for extending the c shim of an existing in-tree/partially-in-tree backend, eg, XPU, those totally out-of-tree backend can generate their c shim normally without this flag.  So here we don't need to specify `if aoti_extend and diapatch_key == "XPU"`

File: caffe2/CMakeLists.txt [hunk 4]
```diff
@@ -325,6 +325,10 @@ set(GENERATED_CXX_TORCH_CUDA
   "${TORCH_SRC_DIR}/csrc/inductor/aoti_torch/generated/c_shim_cuda.cpp"
   )
 
+set(GENERATED_CXX_TORCH_XPU
+  "${TORCH_SRC_DIR}/csrc/inductor/aoti_torch/generated/c_shim_xpu.cpp"
```
Comment (desertfire) [inline caffe2/CMakeLists.txt:329]: Missing `extend`?

Comment (desertfire) [reply caffe2/CMakeLists.txt:329]: NVM: after reading a few more lines, I realized you actually have two parts of c_shim_xpu,

**Review by desertfire — APPROVED** [review]

File: caffe2/CMakeLists.txt (hunk 4 above)
Comment (etaf) [reply caffe2/CMakeLists.txt:329]: @desertfire yes, as I descripted in the PR description above, there is 
```
torch/include/torch/csrc/inductor/aoti_torch/generated/c_shim_xpu.h
torch/include/torch/csrc/inductor/aoti_torch/generated/c_shim_xpu.cpp
torch/include/torch/csrc/inductor/aoti_torch/generated/extend/c_shim_xpu.h
torch/include/torch/csrc/inductor/aoti_torch/generated/extend/c_shim_xpu.cpp
```
The extend part is generated from out-of-tree torch-xpu-ops, which call `python -m torchgen.gen --source-path third_party/torch-xpu-ops/yaml/ --xpu --aoti-extend --update-aoti-c-shim --aoti-insall-dir=torch/include/torch/csrc/inductor/aoti_torch/generated/extend/`, and the generated `extend/c_shim_xpu.cpp` is built  into libtorch_xpu_ops.a. So  `extend/c_shim_xpu.cpp`  is added in cmake file of torch-xpu-ops, not in pytorch cmake file.

**Pull request closed without merging**